# Funciones y paquetes

*Carlos Isaac Zainea Maya*

Python cuenta con una cantidad innumerable de aportes de toda la comunidad dispuestos en archivos (modulos) guardados en algunas carpetas (paquetes). Un elemento importante de estos aportes son las funciones y las clases, de hecho, en una cantidad significativa de ejemplos los desarrolladores buscan una función o clases para apoyar sus creaciones.

En este cuaderno estudiaremos ese componente fundamental de varios bloques de código, *las funciones* y conversaremos acerca de la instalación, inspección y carga de paquetes.

## Funciones

Las funciones son bloques de código en la cuál la máquina espera uno o varios argumentos de entrada. De acuerdo con lo que recibe, se ejecutan unos pasos y finalmente se produce un argumento de salida.

La estructura para definirlas es la siguiente:

>def nombre_de_la_funcion(ENTRADA):

>    Pasos a realizar

>    return SALIDA


Hay una lista de funciones construidas en Python. En este [link](https://docs.python.org/3/library/functions.html), se encuentra un manual de todas las funciones básicas de Python.

Veamos algunas de ellas:


**Ejemplo 1** Funciones predefinidas en Python.

`all()` y `any()` 

In [1]:
y=[4<1,5<6,6<7]
all(y)

False

In [2]:
y=[4>1,5<6,6<7]
all(y)

True

In [3]:
y=[4<1,5<6,6<7]
any(y)

True

In [4]:
y=[4<1,8<6,9<7]
any(y)

False

In [6]:
all((1,1,1))

True

In [7]:
mydict = { 1 : "Orange"}
all(mydict)

True

`dict()`

In [8]:
dict(nombre='Isaac',intereses=['Academicos','Profesionales'])

{'nombre': 'Isaac', 'intereses': ['Academicos', 'Profesionales']}

In [9]:
a = dict(one=1, two=2, three=3)
b = {'one': 1, 'two': 2, 'three': 3}
c = dict(zip(['one', 'two', 'three'], [1, 2, 3]))
d = dict([('two', 2), ('one', 1), ('three', 3)])
e = {'three': 3, 'one': 1, 'two': 2}
print(a == b == c == d == e)
print(e)

True
{'three': 3, 'one': 1, 'two': 2}


In [10]:
list(zip(['one', 'two', 'three'], [1, 2, 3]))

[('one', 1), ('two', 2), ('three', 3)]

In [11]:
a = dict(one={1}, two={2,2}, three={3,3,3})
a

{'one': {1}, 'two': {2}, 'three': {3}}

`abs()`

In [13]:
print(abs(10))
abs(10)==abs(-10)

10


True

`eval()`

In [15]:
x = 6
eval('x**2')

36

`map()`

In [ ]:
x=map(len, {'apple', 'blueberry', 'cherry'}) 
list(x)

`reversed`

In [ ]:
x=[1,2,3]
y=reversed(x)
list(y)

In [ ]:
x="CursoPython"
y=reversed(x)
z=list(y)
print(z)

In [ ]:
y="".join(z)
x=reversed(y)
z=list(x)
z="".join(z)
print(z)

In [16]:
"Cuaderno Python".count("n")

2

In [17]:
"Cuaderno Python".index("a")

2

In [ ]:
Z="Cuaderno de Python hecho en jupyter".split()
Z

In [ ]:
"20%".join(Z)

In [ ]:
"Carta".replace("a","o")

In [ ]:
"Canción".replace("ó","o")

`print()` y `format`

In [ ]:
"Hola {0}, tu saldo es {1:.3f}".format("Juan",321.43212)

In [ ]:
N=input("Tu nombre")
S=input("Tu saldo")
S=float(S)
if S>200000:
    print("Hola {name}, tu saldo es {saldo:.2f}, afortunadamente {name} viviras otra semana y se te dará un premio.".format(name=N,saldo=S))
else:
    print("Hola {name}, tu saldo es {saldo:.2f}, recuerda {name} que hay que ahorrar.".format(name=N,saldo=S))

In [ ]:
print("Hola {0}, tu saldo es {1:9.3f}".format("Pedro",320000.2))

In [ ]:
# integer arguments
print("El número es:{:d}".format(123))

# float arguments
print("El número con dos decimales:{:.2f}".format(123.4567898))

# octal, binary and hexadecimal format
print("bin: {0:b}, oct: {0:o}, hex: {0:x}".format(12))

#exp notation, % notation
print("Exp Notation: {0:.2e},% Notation: {0:.1%}".format(0.12))

In [ ]:

txt = "The universe is {:,} years old."

print(txt.format(13800000000))


In [ ]:
Nombres=["Juan","Pedro","Carlos"]
Saldos=[100,200,300]

N=input("Nombre")
ind=Nombres.index(N)

print("Hola {name}, tu saldo es {saldo:.2f}.".format(name=Nombres[ind],saldo=Saldos[ind]))

`min(), max(), sum() `

In [ ]:
print(min(range(10)))
print(max(range(10)))
print(sum(range(10)))


**Ejemplo 2** También podemos  crear funciones, el siguiente código define una función que recibe un texto y en su parte final pone lo que mide:

In [18]:
def cuantomide(texto):
    n=len(texto)
    print(texto, "tiene",n,"caracteres")
    return n

In [21]:
cuantomide("Universidad Externado")

Universidad Externado tiene 21 caracteres


21

Veamos que dice acerca de la frase "Curso de Python Básico":


In [22]:
cuantomide("Curso de Python Básico")

Curso de Python Básico tiene 22 caracteres


22

Observe que la salida de la función corresponde a la cantidad de caracteres, la frase hace parte de lo que ejecuta la función (en este caso `print`) sin embargo no aporta al resultado.

Algunas funciones útiles:

In [23]:
def soyadulto(x):
  if x < 18:
    return False
  else:
    return True

In [25]:
soyadulto(15)

False

In [26]:
edades = [5, 12, 17, 18, 24, 32]

def soyadulto(x):
  if x < 18:
    return False
  else:
    return True

adultos = filter(soyadulto, edades)

for x in adultos:
  print(x) 

18
24
32


In [28]:
def sumar5ydividirpor10(x):
    y=(x+5)/10
    return y

In [29]:
sumar5ydividirpor10(15)

2.0

# Inspección, Instalación y Carga de Módulos y Paquetes



Los módulos y paquetes de Python son funciones útiles que ya han sido creadas y están a nuestra disposición en la nube. Dependiendo del ambiente de trabajo usado, tenemos diferente cantidad de **paquetes** instalados. 

En la primera instalación de Anaconda y jupyter se instalaron varios paquetes, para verlos usamos la siguiente linea de código:

In [77]:
# Recuerda que # es para usar comentarios
# Versión Actual de Python siendo usada
!python --version
# Función print para dar un salto de línea
#!conda list

Python 3.8.5


¡Aparecen un montón! entre los más importantes están:

* Numpy
* Scipy
* Pandas
* Sympy
* Matplotlib

Dentro de cada uno de estos paquetes se encuentran **modulos**, que juntos forman la librería de módulos disponibles en el ambiente de trabajo. 

Si alguno de los anteriores paquetes no aparece en la lista se puenden instalar escribiendo en una celda de código:

`!pip install <nombre_paquete>`

o

`!conda install <nombre_paquete>`

Uno que siempre falta es sympy, lo instalamos entonces:

In [32]:
!pip install vyper

  Using cached vyper-0.2.8-py3-none-any.whl (219 kB)
  Using cached asttokens-2.0.4-py2.py3-none-any.whl (20 kB)
  Using cached semantic_version-2.8.5-py2.py3-none-any.whl (15 kB)
     |████████████████████████████████| 1.9 MB 5.3 MB/s eta 0:00:01



## Qué es un módulo y como se usa:

Un módulo es un archivo que tiene extensión *.py*, es decir, un archivo de texto que contiene código de python que se puede ejecutar. Un módulo puede definir funciones, clases y variables.

Cabe resaltar lo siguiente: **Python es un lenguaje de programación orientado a objetos**.

Para términos de visualización, imagínese una caja que tiene muchas cajas por dentro y dentro de cada caja existen herramientas de trabajo distintas.

**Ejemplo:**

Mi maleta. Mi maleta tiene dentro lo siguiente:

- Una cartuchera.
- Un computador.
- Un cuaderno.

Los roles serían los siguientes:

- Maleta     $\rightarrow$ Paquete
- Cartuchera $\rightarrow$ Módulo
- Cuaderno   $\rightarrow$ Módulo

Entonces, si queremos usar la cartuchera, debemos primero abrir la maleta y tomarla.

**O sea, del paquete, vamos a usar un módulo**.

Dentro de la cartuchera tengo marcadores, lápices, y borrador.

Estos toman el rol de **atributos**, objetos que hacen tareas específicas.

Por tanto, si quiero usar un borrador, debería hacer lo siguiente

Abrir maleta $\rightarrow$ Tomar cartuchera $\rightarrow$ Abrir Cartuchera $\rightarrow$ Tomar Borrador

En Python, estas acciones se pueden **traducir** a lo siguiente:

```import Maleta as mt```

```mt.Cartuchera.Borrador()```



### Paquete random

Veamos alguna documentación de alguno de los módulos y sus atributos.

Por ejemplo random.

In [41]:
#help("random") # Muestra toda la documentación

import random

?random # Muestra una pequeña documentación

Object `random # Muestra una pequeña documentación` not found.


In [34]:
random.randint(3,15)

6

In [38]:
random.sample(["Pedro","Juan","Carlos"],2)

['Carlos', 'Juan']

In [37]:
dir(random)

['BPF',
 'LOG4',
 'NV_MAGICCONST',
 'RECIP_BPF',
 'Random',
 'SG_MAGICCONST',
 'SystemRandom',
 'TWOPI',
 '_Sequence',
 '_Set',
 '__all__',
 '__builtins__',
 '__cached__',
 '__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__spec__',
 '_accumulate',
 '_acos',
 '_bisect',
 '_ceil',
 '_cos',
 '_e',
 '_exp',
 '_inst',
 '_log',
 '_os',
 '_pi',
 '_random',
 '_repeat',
 '_sha512',
 '_sin',
 '_sqrt',
 '_test',
 '_test_generator',
 '_urandom',
 '_warn',
 'betavariate',
 'choice',
 'choices',
 'expovariate',
 'gammavariate',
 'gauss',
 'getrandbits',
 'getstate',
 'lognormvariate',
 'normalvariate',
 'paretovariate',
 'randint',
 'random',
 'randrange',
 'sample',
 'seed',
 'setstate',
 'shuffle',
 'triangular',
 'uniform',
 'vonmisesvariate',
 'weibullvariate']

In [39]:
random.sample(dir(random),5)

['_warn', 'betavariate', '_os', 'expovariate', 'NV_MAGICCONST']

El comando `dir(random)` nos permite ver los módulos que podemos usar, observe que aparece la función random, tengo que poner primero el nombre del paquete y luego el nombre de la función:

In [42]:
x = random.random()
print("Número aleatorio en [0,1]: ",x)

Número aleatorio en [0,1]:  0.6087761756960691


Otra función en random es `randint`

In [43]:
?random.randint

Signature: random.randint(a, b)
Docstring:
Return random integer in range [a, b], including both end points.
        
File:      ~/anaconda3/lib/python3.8/random.py
Type:      method


In [44]:
y=random.randint(3,6)
print("Número entero aleatorio entre 3 y 6: ",y)

Número entero aleatorio entre 3 y 6:  4


## Alias para módulos

Otra propiedad genial de Python es su capacidad para reducir la cantidad de código escrito usando unos secretos milenarios: aliases.

Por ejemplo, si me llamo Daniel, mi **alias** puede ser Dani.

Entonces, acada vez que me digan Dani, yo miraré en respuesta.

Traducir a Python luce así:

```import modulo as alias```

**Ejemplo:**

In [45]:
import math as m

# Una vez nombrado el alias, siempre se debe referir a ese alias y no al nombre principal

#dir(math) genera error
#dir(m)

In [48]:
dir(m)

['__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__spec__',
 'acos',
 'acosh',
 'asin',
 'asinh',
 'atan',
 'atan2',
 'atanh',
 'ceil',
 'comb',
 'copysign',
 'cos',
 'cosh',
 'degrees',
 'dist',
 'e',
 'erf',
 'erfc',
 'exp',
 'expm1',
 'fabs',
 'factorial',
 'floor',
 'fmod',
 'frexp',
 'fsum',
 'gamma',
 'gcd',
 'hypot',
 'inf',
 'isclose',
 'isfinite',
 'isinf',
 'isnan',
 'isqrt',
 'ldexp',
 'lgamma',
 'log',
 'log10',
 'log1p',
 'log2',
 'modf',
 'nan',
 'perm',
 'pi',
 'pow',
 'prod',
 'radians',
 'remainder',
 'sin',
 'sinh',
 'sqrt',
 'tan',
 'tanh',
 'tau',
 'trunc']

### Aplicación:

Una pequeña aplicación de Matemáticas:

**Área de una circunferencia de radio r:**

$A = \pi r²$

In [50]:
m.pi

3.141592653589793

In [51]:
# Aplicación: Área de un círculo de radio 10 cm 

r = 10;          # Declarar el radio
z = m.pi*(r**2)  # Escribir la fórmula matemática
print("El Área del círculo es: ",z,"cm²")

El Área del círculo es:  314.1592653589793 cm²


Como pueden observar, los módulos son el **alma de Python**.



In [52]:
from math import pi
r=2
A = pi*r**2

print(A)

12.566370614359172


In [53]:
m.pi

3.141592653589793

In [54]:
pi

3.141592653589793

In [55]:
import math
dir(math)

['__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__spec__',
 'acos',
 'acosh',
 'asin',
 'asinh',
 'atan',
 'atan2',
 'atanh',
 'ceil',
 'comb',
 'copysign',
 'cos',
 'cosh',
 'degrees',
 'dist',
 'e',
 'erf',
 'erfc',
 'exp',
 'expm1',
 'fabs',
 'factorial',
 'floor',
 'fmod',
 'frexp',
 'fsum',
 'gamma',
 'gcd',
 'hypot',
 'inf',
 'isclose',
 'isfinite',
 'isinf',
 'isnan',
 'isqrt',
 'ldexp',
 'lgamma',
 'log',
 'log10',
 'log1p',
 'log2',
 'modf',
 'nan',
 'perm',
 'pi',
 'pow',
 'prod',
 'radians',
 'remainder',
 'sin',
 'sinh',
 'sqrt',
 'tan',
 'tanh',
 'tau',
 'trunc']

In [56]:
??math.pi

Type:        float
String form: 3.141592653589793
Docstring:   Convert a string or number to a floating point number, if possible.


In [57]:
??m.lgamma

Signature: m.lgamma(x, /)
Docstring: Natural logarithm of absolute value of Gamma function at x.
Type:      builtin_function_or_method


## Paquete numpy

NumPy es un paquete de Python para hacer cálculos numéricos, su nombre proviene de la abreviación de “Numerical Python”, es la librería principal para la computación científica y contiene las funciones matemáticas más famosas, tambien algunas constantes de interes: 


In [58]:
import numpy as np #np es el alias más conocido de numpy

Para usar cualquier función de numpy tenemos que escribir

`np.<función>`

A continuación calculamos:

* $\pi=3.14159\cdots$

* $|-10|=10$

* $e^1$


In [59]:
np.pi

3.141592653589793

In [60]:
np.abs(-10)

10

In [61]:
np.exp(1)

2.718281828459045

In [62]:
dir(np)

['ALLOW_THREADS',
 'AxisError',
 'BUFSIZE',
 'CLIP',
 'ComplexWarning',
 'DataSource',
 'ERR_CALL',
 'ERR_DEFAULT',
 'ERR_IGNORE',
 'ERR_LOG',
 'ERR_PRINT',
 'ERR_RAISE',
 'ERR_WARN',
 'FLOATING_POINT_SUPPORT',
 'FPE_DIVIDEBYZERO',
 'FPE_INVALID',
 'FPE_OVERFLOW',
 'FPE_UNDERFLOW',
 'False_',
 'Inf',
 'Infinity',
 'MAXDIMS',
 'MAY_SHARE_BOUNDS',
 'MAY_SHARE_EXACT',
 'MachAr',
 'ModuleDeprecationWarning',
 'NAN',
 'NINF',
 'NZERO',
 'NaN',
 'PINF',
 'PZERO',
 'RAISE',
 'RankWarning',
 'SHIFT_DIVIDEBYZERO',
 'SHIFT_INVALID',
 'SHIFT_OVERFLOW',
 'SHIFT_UNDERFLOW',
 'ScalarType',
 'Tester',
 'TooHardError',
 'True_',
 'UFUNC_BUFSIZE_DEFAULT',
 'UFUNC_PYVALS_NAME',
 'VisibleDeprecationWarning',
 'WRAP',
 '_NoValue',
 '_UFUNC_API',
 '__NUMPY_SETUP__',
 '__all__',
 '__builtins__',
 '__cached__',
 '__config__',
 '__dir__',
 '__doc__',
 '__file__',
 '__getattr__',
 '__git_revision__',
 '__loader__',
 '__mkl_version__',
 '__name__',
 '__package__',
 '__path__',
 '__spec__',
 '__version__',
 '_ad

In [63]:
np.mean([1,2,3])

2.0

In [65]:
dir(np.linalg)

['LinAlgError',
 '__builtins__',
 '__cached__',
 '__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__path__',
 '__spec__',
 '_umath_linalg',
 'cholesky',
 'cond',
 'det',
 'eig',
 'eigh',
 'eigvals',
 'eigvalsh',
 'inv',
 'lapack_lite',
 'linalg',
 'lstsq',
 'matrix_power',
 'matrix_rank',
 'multi_dot',
 'norm',
 'pinv',
 'qr',
 'slogdet',
 'solve',
 'svd',
 'tensorinv',
 'tensorsolve',
 'test']

In [69]:
np.linalg.svd(np.array([[2,4],[1,2]]))

(array([[-0.89442719, -0.4472136 ],
        [-0.4472136 ,  0.89442719]]),
 array([5.00000000e+00, 1.39218791e-17]),
 array([[-0.4472136 , -0.89442719],
        [ 0.89442719, -0.4472136 ]]))

In [70]:
np.sin(32)

0.5514266812416906

In [71]:
V1=np.array([1,2])

In [72]:
V2=np.array([2,3])

In [73]:
V1+V2

array([3, 5])

In [74]:
M=np.array([[1,2],[2,3]])

In [75]:
np.linalg.eig(M)

(array([-0.23606798,  4.23606798]),
 array([[-0.85065081, -0.52573111],
        [ 0.52573111, -0.85065081]]))

In [76]:
?np.linalg.eig

Signature: np.linalg.eig(a)
Docstring:
Compute the eigenvalues and right eigenvectors of a square array.

Parameters
----------
a : (..., M, M) array
    Matrices for which the eigenvalues and right eigenvectors will
    be computed

Returns
-------
w : (..., M) array
    The eigenvalues, each repeated according to its multiplicity.
    The eigenvalues are not necessarily ordered. The resulting
    array will be of complex type, unless the imaginary part is
    zero in which case it will be cast to a real type. When `a`
    is real the resulting eigenvalues will be real (0 imaginary
    part) or occur in conjugate pairs

v : (..., M, M) array
    The normalized (unit "length") eigenvectors, such that the
    column ``v[:,i]`` is the eigenvector corresponding to the
    eigenvalue ``w[i]``.

Raises
------
LinAlgError
    If the eigenvalue computation does not converge.

See Also
--------
eigvals : eigenvalues of a non-symmetric array.
eigh : eigenvalues and eigenvectors of a real symmet

A lo largo del curso usaremos varias funciones alojadas en paquetes. Cada vez que sea necesario encontraran una pequeña ayuda para usar dicha función.